# Combining Data

Practice combining data from two different data sets. In the same folder as this Jupyter notebook, there are two csv files:
* rural_population_percent.csv
* electricity_access_percent.csv

They both come from the World Bank Indicators data. 
* https://data.worldbank.org/indicator/SP.RUR.TOTL.ZS
* https://data.worldbank.org/indicator/EG.ELC.ACCS.ZS

The rural populaton data represents the percent of a country's population that is rural over time. The electricity access data shows the percentage of people with access to electricity.

In this exercise, you will combine these two data sets together into one pandas data frame.

# Exercise 1

Combine the two data sets using the [pandas concat method](https://pandas.pydata.org/pandas-docs/stable/merging.html). In other words, find the union of the two data sets.

In [70]:
# TODO: import the pandas library
import pandas as pd

# TODO: read in each csv file into a separate variable
# HINT: remember from the Extract material that these csv file have some formatting issues
# HINT: The file paths are 'rural_population_percent.csv' and 'electricity_access_percent.csv'
df_rural = pd.read_csv('rural_population_percent.csv', skiprows=range(0,4))
df_electricity = pd.read_csv('electricity_access_percent.csv', skiprows=range(0,4))

# TODO: remove the 'Unnamed:62' column from each data set
df_rural = df_rural.drop(['Unnamed: 62'], axis=1)
df_electricity = df_electricity.drop(['Unnamed: 62'], axis=1)

# TODO: combine the two data sets together using the concat method
# In other words, all of the rows of df_rural will come first
# followed by all the rows in df_electricity. This is possible to do
# because they both have the same column names.
df_combined = pd.concat([df_rural, df_electricity], axis=1)


In [71]:
df_combined.head()

Country Name Country Code                            Indicator Name  \
0        Aruba          ABW  Rural population (% of total population)   
1  Afghanistan          AFG  Rural population (% of total population)   
2       Angola          AGO  Rural population (% of total population)   
3      Albania          ALB  Rural population (% of total population)   
4      Andorra          AND  Rural population (% of total population)   

   Indicator Code    1960    1961    1962    1963    1964    1965  ...   \
0  SP.RUR.TOTL.ZS  49.224  49.239  49.254  49.270  49.285  49.300  ...    
1  SP.RUR.TOTL.ZS  91.779  91.492  91.195  90.890  90.574  90.250  ...    
2  SP.RUR.TOTL.ZS  89.565  89.202  88.796  88.376  87.942  87.496  ...    
3  SP.RUR.TOTL.ZS  69.295  69.057  68.985  68.914  68.842  68.770  ...    
4  SP.RUR.TOTL.ZS  41.550  39.017  36.538  34.128  31.795  29.555  ...    

         2008        2009        2010        2011        2012        2013  \
0   93.086166   93.354546   93.356292   93.942375   94.255814   94.578262   
1   42.400000   44.854885   42.700000   43.222019   69.100000   67.259552   
2   31.268013   32.382469   33.514950   34.600000   35.821964   36.990490   
3  100.000000  100.000000  100.000000  100.000000  100.000000  100.000000   
4  100.000000  100.000000  100.000000  100.000000  100.000000  100.000000   

         2014        2015        2016  2017  
0   94.906723   95.238182   95.570145   NaN  
1   89.500000   71.500000   84.137138   NaN  
2   32.000000   42.000000   40.520607   NaN  
3  100.000000  100.000000  100.000000   NaN  
4  100.000000  100.000000  100.000000   NaN  

[5 rows x 124 columns]

# Exercise 2 (Challenge)

This exercise is more challenging.

Combine the two datas in the csv file together so that the output looks like the following:

|Country Name|Country Code|Year|Rural_Value|Electricity_Value|
|--|--|--|--|--|
|Aruba|ABW|1960|49.224|49.239|

... etc.

Order the results in the dataframe by country and then by year

Here are a few pandas methods that should be helpful:
* [melt](http://pandas.pydata.org/pandas-docs/stable/generated/pandas.melt.html)
* [drop](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.drop.html)
* [merge](https://pandas.pydata.org/pandas-docs/version/0.23/generated/pandas.DataFrame.merge.html)
* [sort_values](https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.sort_values.html)

HINT: You can use country name, country code, and the year as common keys between the data sets

In [74]:
# TODO: merge the data sets together according to the instructions. First, use the 
# melt method to change the formatting of each data frame so that it looks like this:
# Country Name, Country Code, Year, Rural Value
# Country Name, Country Code, Year, Electricity Value

df_rural_melt = pd.melt(df_rural, id_vars=['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code'], var_name = 'Year', value_name='Rural_Value')
df_electricity_melt = pd.melt(df_electricity, id_vars=['Country Name', 'Country Code', 'Indicator Name', 'Indicator Code'], var_name = 'Year', value_name='Electricity_Value')

# TODO: drop any columns from the data frames that aren't needed
df_rural_clean = df_rural_melt.drop(['Indicator Name', 'Indicator Code'], axis=1)
df_electricity_clean = df_electricity_melt.drop(['Indicator Name', 'Indicator Code'], axis=1)

# TODO: merge the data frames together based on their common columns
# in this case, the common columns are Country Name, Country Code, and Year
df_combined = df_electricity_clean.merge(df_rural_clean, how='inner')

# TODO: sort the results by country and then by year
df_combined = df_combined.sort_values(by=['Country Name', 'Year'])
df_combined = df_combined.reset_index(drop=True)


In [76]:
df_combined.head(100)

Country Name Country Code  Year  Electricity_Value  Rural_Value
0   Afghanistan          AFG  1960                NaN       91.779
1   Afghanistan          AFG  1961                NaN       91.492
2   Afghanistan          AFG  1962                NaN       91.195
3   Afghanistan          AFG  1963                NaN       90.890
4   Afghanistan          AFG  1964                NaN       90.574
5   Afghanistan          AFG  1965                NaN       90.250
6   Afghanistan          AFG  1966                NaN       89.915
7   Afghanistan          AFG  1967                NaN       89.570
8   Afghanistan          AFG  1968                NaN       89.214
9   Afghanistan          AFG  1969                NaN       88.848
10  Afghanistan          AFG  1970                NaN       88.471
11  Afghanistan          AFG  1971                NaN       88.083
12  Afghanistan          AFG  1972                NaN       87.684
13  Afghanistan          AFG  1973                NaN       87.274
14  Afghanistan          AFG  1974                NaN       86.851
15  Afghanistan          AFG  1975                NaN       86.417
16  Afghanistan          AFG  1976                NaN       85.971
17  Afghanistan          AFG  1977                NaN       85.513
18  Afghanistan          AFG  1978                NaN       85.042
19  Afghanistan          AFG  1979                NaN       84.565
20  Afghanistan          AFG  1980                NaN       84.319
21  Afghanistan          AFG  1981                NaN       84.070
22  Afghanistan          AFG  1982                NaN       83.818
23  Afghanistan          AFG  1983                NaN       83.563
24  Afghanistan          AFG  1984                NaN       83.304
25  Afghanistan          AFG  1985                NaN       83.042
26  Afghanistan          AFG  1986                NaN       82.777
27  Afghanistan          AFG  1987                NaN       82.509
28  Afghanistan          AFG  1988                NaN       82.237
29  Afghanistan          AFG  1989                NaN       81.962
..          ...          ...   ...                ...          ...
70      Albania          ALB  1972                NaN       67.873
71      Albania          ALB  1973                NaN       67.679
72      Albania          ALB  1974                NaN       67.484
73      Albania          ALB  1975                NaN       67.288
74      Albania          ALB  1976                NaN       67.092
75      Albania          ALB  1977                NaN       66.895
76      Albania          ALB  1978                NaN       66.698
77      Albania          ALB  1979                NaN       66.500
78      Albania          ALB  1980                NaN       66.238
79      Albania          ALB  1981                NaN       65.976
80      Albania          ALB  1982                NaN       65.713
81      Albania          ALB  1983                NaN       65.448
82      Albania          ALB  1984                NaN       65.183
83      Albania          ALB  1985                NaN       64.917
84      Albania          ALB  1986                NaN       64.650
85      Albania          ALB  1987                NaN       64.381
86      Albania          ALB  1988                NaN       64.112
87      Albania          ALB  1989                NaN       63.842
88      Albania          ALB  1990              100.0       63.572
89      Albania          ALB  1991              100.0       63.300
90      Albania          ALB  1992              100.0       62.751
91      Albania          ALB  1993              100.0       62.201
92      Albania          ALB  1994              100.0       61.646
93      Albania          ALB  1995              100.0       61.089
94      Albania          ALB  1996              100.0       60.527
95      Albania          ALB  1997              100.0       59.965
96      Albania          ALB  1998              100.0       59.399
97      Albania          ALB  1999              10